### ISSUE TITLE
Is it possible to retrieve the attention weights of a specific node?

### ISSUE THREAD

**Issue by sgdantas** [dmlc/dgl#608]
Hey!
I was wondering if it's possible to retrieve the attention weights of a specific node. By printing the alpha dimension I see that the attention is batched (Number of nodes with same degree x degree x 1)
```
def reduce_func(self, nodes):
        # reduce UDF for equation (3) & (4)
        # equation (3)
        alpha = F.softmax(nodes.mailbox['e'], dim=1)
        print(alpha.shape)
        # equation (4)
        h = torch.sum(alpha * nodes.mailbox['z'], dim=1)
        return {'h': h}
```

However, a node might be connected to other nodes with different degrees. I wanted to retrieve the attention that other nodes pay to a specific node, is that possible?

Thanks!

**Comment by mufeili** [comment]
It's possible. Take [our PyTorch GAT Implementation](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/gat.py) as an example. The edge attentions are stored in [`g.edata['a_drop']`](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/gat.py#L88). We can fetch the attentions as follows:

```python
from scipy.sparse import lil_matrix

def preprocess_attention(edge_atten, g, to_normalize=True):
    """Organize attentions in the form of dense adjacency
    matrices from attention on edges.

    Parameters
    ----------
    edge_atten : numpy.array of shape (# edges, # heads, 1)
        Un-normalized attention on edges.
    g : dgl.DGLGraph.
    to_normalize : bool
        Whether to normalize attention values over incoming
        edges for each node.
    """
    n_nodes = g.number_of_nodes()
    num_heads = edge_atten.shape[1]
    all_head_A = [lil_matrix((n_nodes, n_nodes)) for _ in range(num_heads)]
    for i in range(n_nodes):
        predecessors = list(g.predecessors(i))
        edges_id = g.edge_ids(predecessors, i)
        for j in range(num_heads):
            all_head_A[j][i, predecessors] = edge_atten[edges_id, j, 0].data.cpu().numpy()
    if to_normalize:
        for j in range(num_heads):
            all_head_A[j] = normalize(all_head_A[j], norm='l1').tocsr()
    return all_head_A

# Take the attention from one layer as an example
A = self.g.edata['a_drop']                             # num_edges x num_heads x 1
A = preprocess_attention(A, self.g)     
```

**Comment by sgdantas** [comment]
Thank you, it works!

**Comment by mufeili** [comment]
Glad to know that @sgdantas F :). Let me know if you have any follow up questions. Also for questions like this we encourage users to post on our [discussion forum](https://discuss.dgl.ai/) so that more users can be benefited.

**Comment by mastreips** [comment]
Hey - the code has been updated since the solution posted above. A = self.g.edata['a_drop'] is gone.  I am able to get the 'attn_l' tensors but there are only 64 of them [1,8,8] tensor for 8 node, 8 head architecture.  My model as 9752 edges, so how would I get the attention values for all of them?  Thanks!

**Comment by yzh119** [comment]
@mastreips take a look at our [edge_softmax](https://docs.dgl.ai/api/python/nn.pytorch.html#module-dgl.nn.pytorch.softmax) api and see if it works for you.
